# Lab 0: Intro to Python, Numpy, and Numeric Types
---


## Lab Objectives

The goals for this lab are to 

- Understand how to import a module, namely numpy. 
- How to work with different numeric types and storage containers.
- Basic manipulation of matrices and vectors

<span style="color:red">Instructions: run through the entire notebook. There are questions at the end of each section that you need to answer/write code for. Submit the completed notebook as your report. </span>

---
## (0) Introduction

#### What is Python?
__Short answer__: It is an interpreted, high level, dynamically typed, language supporting multiple features such as object oriented and functional programming.<br>
__Shorter answer__: It is a programming language that was designed to be easy to use.

#### What is Numpy?
A package for scientific computing in Python. It is a library that provides multidimensional arrays, pseuso-random number generation, basic linear algebra and other basic routines needed for data science. <br>
https://docs.scipy.org/doc/numpy/

#### What is Jupyter Notebook/Lab?
Jupyter Notebook/Lab is a web application that allow us to interactively run code and analyze intermediate results. 
 
For this class we will write all our code in Jupyter notebooks in Python. The objective of the labs is to familiarize ourselves with programming in Python using Jupyter notebooks for scientific computing. In our notebooks, there will be two main types of cells, 1) Markdown and 2) Code. Markdown cells are like this cell and provide an environment for writing formatted text. Code cells will contain the code that is ultimately executed. To process a cell, we press `shift enter`.



---
## (1) Importing Modules

First, we must load the modules we would like to use. A module can be as a simple as a text file containing some frequently used code. We use the `import` command to import a particular module, and then use `as` to provide a shorter name to access it if we desire. Using `as` is not necessary.
<br>

For example python provides a module called math. We may 'import math as mt' if we would like by typing the following code.

```python
import math as mt
```

Once we would like to execute the code in a cell, we type `shift enter`. 


After executing the code, anything that the imported module provides will be accessed through typing 

`{module name}.{name of submodule/function}`. 

We could have used any other name besides `mt`, but for convenience chose `mt`. Execute the below code block.

In [1]:
## Here we import the math module as mt to save use from typing an extra 2 letters.
import math as mt

## Let us calculate the square root of 16
mt.sqrt(16)

4.0

## <span style="color:red"> (1) Questions</span>

1. Below create a new code cell below.
2. Import the 'numpy' library as 'np'.
3. Use the numpy provided function `sqrt` to calculate the square root of 25.
4. Execute the code in the cell.

## <span style="color:red"> (1) Answers</span>

#### How to Install numpy
Run the following piece of code inside the Jupyter notebook:

```python
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install numpy
```

In [2]:
#importing numpy
import numpy as np

#printing answer.
np.sqrt(25)

5.0

---
## (2) Numeric Types and Floating Point Arithmetic



Imagine we perform an experiment, the results or observations of our experiment can be of many different types;

- Flip a coin
    - Heads or Tails -> True or False (Boolean)
    
- Measure a Heartbeat
    - $60.2$ beats per minute -> $0.0 - 500.0$ (Float)
    
- Demodulate a RF signal
    - $1.0 - 1.0\cdot j$ (I and Q) -> Complex Numbers in a disk (Complex)
    

Numbers are approximated on a computer and we are not truely able to represent all real numbers. In particular we have limited precision. Floating point numbers are represented as a sum of base 2 fractions times an exponent, as below. 
$$ (-1)^{sign}(1 + \sum_{i=1}^{m} \frac{\alpha_i}{2^i} )\cdot 2^{\beta} $$ 

This representation may lead to unexpected behavior. We can see a simple example below where a sum does not equal what we expect it to equal.

In [3]:
# Here we test whether summing three floating point numbers is equal to what we think it should equal.
0.1 + 0.1 + 0.1 == 0.3

False

We see below the numerical difference.

In [4]:
(0.1 + 0.1 + 0.1) - 0.3

5.551115123125783e-17

Understanding the conventions for representing numbers is an essential part of scientific programming. We must be aware of these limitations, and account for them so that our code will have the correct semantics. In particular, we may consider two floating point numbers the same if they agree up to a certain number of digits instead of above. 

To represent data types in numpy, we have numeric types, which are boolean, int, float, complex. Our goal is to use the minimal representation to store our data. For int, float, and complex we may choose a size of the representation, such as 32-bit and 64-bit values, although there are more possible values. To see a complete list go to <br>
https://numpy.org/devdocs/user/basics.types.html


To specify the type of a number or variable, we use <br>
`np.(variable type)` <br>
Types will be important since we need to understand precisely what calculations we are performing, and why we may end up with nans and infs.

<br>
<div class="alert alert-block alert-info">
<b> Python Types:</b> Python provides an integer type where the size of the representation can grow. We will not cover it here but more can be read at  <a href="http://site.com/product/product-link#modal">
    http://www.laurentluce.com/posts/python-integer-objects-implementation/
</div>
    
Below we create a variable `x` and initialize it with a floating point number then convert it to an integer.

In [5]:
import numpy as np

x=np.float32(1.349) #set a variable to a 32-bit floating point value
print("x has type ",type(x)," and value ",x) # print the variable x and its type

x=np.int32(x) # convert x to a 32-bit integer
print("x has type ",type(x)," and value ",x) # print the variable x and its type


x has type  <class 'numpy.float32'>  and value  1.349
x has type  <class 'numpy.int32'>  and value  1


We recall the example several cells above and test how specifying the number of bits for a floating point affect checking whether two numbers are equal.

In [6]:
np.float64(.1) + np.float64(.1) + np.float64(.1) == np.float64(.3)

False

In [7]:
np.float32(.1) + np.float32(.1) + np.float32(.1) == np.float32(.3)

True

We see above that specifying the number of bits for a floating point representation may affect the outcome of our code.

Below we create a variable `y` and initialize it with a larger floating point number then attempt to convert it to a 32-bit integer.


In [8]:

y=2147483648.0
print("y has type ",type(y)," and value ",y) # print the variable x and its type

y=int(y) # convert x to a 32-bit integer
print("y has type ",type(y)," and value ",y) # print the variable x and its type


# attmmpt to convert y to a 32-bit integer but fail.
y=np.int32(y) # convert x to a 32-bit integer
print("y has type ",type(y)," and value ",y) # print the variable x and its type

y has type  <class 'float'>  and value  2147483648.0
y has type  <class 'int'>  and value  2147483648
y has type  <class 'numpy.int32'>  and value  -2147483648


As we notice above, we must be careful with the types of variables we use.

**Below we output the possible values of various size integers.**

In [9]:
print(np.iinfo(np.int8)) # Bounds of a 8-bit integer 

print(np.iinfo(np.int32) )# Bounds of a 32-bit integer

print(np.iinfo(np.int64)) # Bounds of a 64-bit integer

Machine parameters for int8
---------------------------------------------------------------
min = -128
max = 127
---------------------------------------------------------------

Machine parameters for int32
---------------------------------------------------------------
min = -2147483648
max = 2147483647
---------------------------------------------------------------

Machine parameters for int64
---------------------------------------------------------------
min = -9223372036854775808
max = 9223372036854775807
---------------------------------------------------------------



**Below we output the possible values of various size floats, and associated constraints.**



In [10]:
print(np.finfo(np.float32) )# Bounds of a 32-bit float

print(np.finfo(np.float64)) # Bounds of a 64-bit float

Machine parameters for float32
---------------------------------------------------------------
precision =   6   resolution = 1.0000000e-06
machep =    -23   eps =        1.1920929e-07
negep =     -24   epsneg =     5.9604645e-08
minexp =   -126   tiny =       1.1754944e-38
maxexp =    128   max =        3.4028235e+38
nexp =        8   min =        -max
---------------------------------------------------------------

Machine parameters for float64
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
---------------------------------------------------------------



#### To determine the type of a number or variable, we may call the `type` function.

In [11]:
print("Here we print the type of 1  :",type(1))


print("Here we print the type of 1.0 :",type(1.0))


print("Here we print the type of a complex number, 1.0+1j  :",type(1.0+1j*1.0))


Here we print the type of 1  : <class 'int'>
Here we print the type of 1.0 : <class 'float'>
Here we print the type of a complex number, 1.0+1j  : <class 'complex'>


## <span style="color:red"> (2) Questions</span>

1. Below create a new code cell below.
2. Come up with another example of a sum of 64 bit floats that do not sum to what we think they should.
3. What numpy datatype, from this link https://numpy.org/devdocs/user/basics.types.html , would you choose to represent data from collecting;  
    1. attendance of this class
    2. heartbeat measurements
    3. number of votes for a United States of America presidential candidate

    In the cell you create below, write one example for each. For example it we wanted to capture whether a heads or tails happened in a coin flip, we would type 
```python 
np.boolean(True)    
```
    

## <span style="color:red"> (2) Answers</span>

2)

To have a better understanding of why the equality below is false; we need to first understand how we represent 0.1 and 0.2 in 64 bit following the IEEE754 standard. Please click [here](https://medium.com/better-programming/why-is-0-1-0-2-not-equal-to-0-3-in-most-programming-languages-99432310d476)
to understand how to represent floating binary points. 
 
The idea is that 0.1 in binary is 1.10011...100 and 0.2 in binary is 1.0011...101; if you add these 2 numbers in binary it would give you 0.11111...100 which is 0.300..004.

Thus, 0.1 + 0.2 != 0.3 in binary.

In [12]:
np.float64(.1) + np.float64(.2) == np.float64(.3)

False

In [13]:
# Why is wrong? because it doesn't equal .3
print(np.float64(.1) + np.float64(.2)) 

0.30000000000000004


3)
A. attendance of this class 
 - uint8 
     - Unsigned integer (0 to 255)
     - The maximum number of attendance for this class will be the number of students in this class which is 22. Thus we do not need a datatype that has a lot of integers. Since the minimum is int8; the available datatype is uint8 and int8.
     - Since the number of students will never be a negative number; I decided to choose uint8 because it has a value from 0 to 255; in contrast to int that has -128 to 127).

In [14]:
# We are importing again in case this is the first cell you run.
import numpy as np
# Assuming there are 22 students in a class.
np.uint8(22)

22

B. heartbeat measurements     
 - uint8 
     - Unsigned integer (0 to 255)
     - The maximum number of bpm is 200. Thus we do not need a datatype that has a lot of integers. Since the minimum is int8; the available datatype is uint8 and int8.
     - Since bpm will never be a negative number; I decided to choose uint8 because it has a value from 0 to 255; in contrast to int that has -128 to 127).
 - uint16
     - Unsigned integer (0 to 65535)
     - However; if we are measuring bpm of animals; we probably need more space to store our numbers as it might go over 255. Thus we might need uint16.

In [15]:
# We are importing again in case this is the first cell you run.
import numpy as np
# Assuming a human heartbeat
print("Human:", np.uint8(169))
# Assuming an animal heartbeat
print("Animal:", np.uint16(1000))

Human: 169
Animal: 1000


C. number of votes for a United States of America presidential candidate 
 - np.uint32 
     - Unsigned integer (0 to 4294967295)
     - Since votes are only positive, we should use uint as it only has a positive number.
     - Also, we have around 300 million people in the USA; thus, we need a data type that can store at least 300 million; thus, uint32 is the smallest data type that can store this.

In [16]:
# We are importing again in case this is the first cell you run.
import numpy as np
# Number of votes for 2012 election
np.uint32(235248000)

235248000

---
## (3) Storing Numeric Types: Tuples, Lists, and Arrays

If we would like to store multiple observations, e.g., a collection of strings, or a matrix, we will use lists, tuples, and arrays.
Lists, tuples, and arrays have different use cases. <br>

- **Tuples** are immutable, i.e., they can not be changed, and frequently are used to specify the dimensions of an array. Their main advantages are fast access and the fact they can not be changed. They are constructed by writing 
```python 
b=(1,2,3) 
```
<br>
- **Lists** allow us flexibility in storing multiple types in a single container. They also allow us to not specify a size and concatenate elements. Due to their flexibility a time penalty is incurred when trying to access and perform operations. They are declared as follows;
```python 
a=["horse","pig","chicken"] 
```
<br>
- **Arrays** can be accessed fast, but may only store one data type. When we use arrays we will use the implementation provided in numpy. 
```python 
c=np.array([1.0,2.0,3.0,4.0])
```
<br>


### (3.1) Working with Tuples

In this section, we will go over some common tuple properties and operations.


In [17]:
## To create a tuple 
nums1=(1,2,3,4)
nums2=(5,6,7,8,9)

## To create a tuple with one element place a comma after the entry
nums3=(10,) 

## We can access elements of a tuple by using brackets
print("The first entry of nums1 is ",nums1[0])

## We can create a new tuple by joining tuples
nums = nums1 + nums2 + nums3
print("The resulting tuple is ",nums)


The first entry of nums1 is  1
The resulting tuple is  (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)


### (3.2) Working with Lists

In this section, we will go over some common list properties and operations.

Below we create a list of pokemon, calculate the length, and print it out.

In [18]:
pokemon=[ "charizard", "lucario", "eevee","pikachu","mewtwo","mew","bulbasaur","garchomp"]

## we can check on the length of the list using "len".
print("The length of the list is ",len(pokemon),". \n")

## We can use a "for loop" to print out the elements.
print("The elements of pokemon are; ")
for p in pokemon:
    print(p)
    
## here we append another pokemon to the list
pokemon.append("arcanine")

## We confirm that that the last element of the list is "arcanine"
print("The last element of the list, pokemon, is",pokemon[-1])


The length of the list is  8 . 

The elements of pokemon are; 
charizard
lucario
eevee
pikachu
mewtwo
mew
bulbasaur
garchomp
The last element of the list, pokemon, is arcanine


Below we use list comprehension to create a list of squared integers.

In [19]:
## First we create list containing the squares of integers up to 10.
## We use another data type called "range"
lsi=[(x+1)*(x+1) for x in range(10) ]
print(lsi)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


### (3.3) Working with Numpy Arrays 

In this section, we will go over arrays. Lists, ranges, and tuples are provided by python. Python does not provide arrays, and we will use the implementation provided by numpy. Recall, we have already imported `numpy` as `np`. To access any properties about numpy arrays, we will use `np.{ }`. 

Numpy arrays can be initialized in many ways. If the datatype is not specified, numpy will try to find the smallest size datatype that can hold the numeric type provided. Below we use several different methods to construct arrays.


In [20]:
## First we convert a list in python to an array without specifying the datatype
a=np.array([0 , 1])

## Here we did not specify the datatype so numpy used numpy.int32
print("The datatype of the elements of a is  ",type(a[0]), " and has size ", a[0].itemsize, " bytes.")

## Now we provide the datatype
a=np.array([0 , 1],dtype=np.uint8)

## Observe the new datatype
print("The datatype of the elements of a is  ",type(a[0]), " and has size ",  a[0].itemsize, " byte.")


The datatype of the elements of a is   <class 'numpy.int64'>  and has size  8  bytes.
The datatype of the elements of a is   <class 'numpy.uint8'>  and has size  1  byte.


We may also initialize arrays with default values, such as zeros or ones. We will need to provide a **shape** for the array in the form of a tuple. The shape tuple can take any number of coordinates. Suppose the shape tuple has one coordinate---we are talking of a vector (which is a 1-d array). If the shape tuple has two coordinates, we are talking of a matrix (a 2-d array as far as python is concerned). It is quite possible you have 3-d, 4-d and higher arrays, but in this course, we will primarily be concerned with 1- and 2-d arrays (and occasionally 3-d arrays when organizing images for example).

For example, let us consider a vector (1-d array) with $n$ elements.
$$a=\begin{bmatrix}a_0\\ a_1 \\ \vdots \\ a_{n-1} \end{bmatrix}$$
Each element is accessed with one subscript, e.g., $a[2]=a_2$ , starting at $0$ . We can initialize a 1-d array of shape $(n,)$ with zeros by writing.
```python
a=np.zeros((n,),dtype=np.float32)
```


<br>
A two dimensional array, e.g., a matrix, will have the form.
$$a=\begin{bmatrix}a_{0,0} & a_{0,1} & \dots a_{0,m-1}  \\ \vdots & \dots &  \vdots \\ a_{n-1,0} & a_{0,1} & \dots a_{n-1,m-1} \end{bmatrix}$$

We can initialize a 2-d array of shape $(n,m)$ with zeros by writing.
 
```python
a=np.zeros((n,m),dtype=np.float32)
```

In the above examples, the arrays had one or two dimensions. We may create arbitrary dimensions for an array.

In [21]:
## here we create a 3-dimensional array with shape (2,3,4) This means 
## 2 different 3x4 matrices, or 2 matrices each with 3 rows and 4 columns
import numpy as np

a=np.zeros((2,3,4),dtype=np.float32)

print("a=",a)

## we can access the shape using the attribute "shape" on the ndarray object (here, a)
print("The shape of the 3-dimensional array a is ",a.shape)

## You could also access the shape by calling the function shape with argument a
np.shape(a)

## You could also fix the second argument and access the resultant 2x4 matrix. The :
## denotes "run through all values this index can take". 
print("a[:,0,:]=", a[:,0,:])




a= [[[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]

 [[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]]
The shape of the 3-dimensional array a is  (2, 3, 4)
a[:,0,:]= [[0. 0. 0. 0.]
 [0. 0. 0. 0.]]


Now we will show multiple ways to iterate over a multidimensional array.

In [22]:

## we may fill a with ascending numbers by using the following comprehension
count=0
for i in range(a.shape[0]):
    for j in range(a.shape[1]):
        for k in range(a.shape[2]):
            a[i,j,k]=count
            count+=1
## now we see how a was filled
print(a)

## You would access the first matrix using a[0]. 
print("a[0]="+'\n', a[0])

## You could also fix the second argument and access the resultant 2x4 matrix. The :
## denotes "run through all values this index can take". 
print("a[:,0,:]=",'\n', a[:,0,:])

## Suppose you wanted all locations that were < 10
print('Locations where the entry is <10','\n',np.where(a<10))

## The output may be tricky to parse. But this is a list of all locations that satisfy 
## entry < 10. The three arrays are along the three indices of a.
## Each position where the entry is < 10 is obtained by picking one entry from the 
## first array, and the corresponding elements from later arrays. For example, 
## the first position we choose the first element of each of the three arrays
## to obtain (0,0,0). The second position that satisfies entry < 10 is (0,0,1),
## next one is (0,0,2), (0,0,3), (0,1,0) and so on.

## If you would like to print out the entries where a<10, you would use:

print('Entries of a that are < 10', '\n', a[np.where(a<10)])

## Note that the output above is an array. 

## You can change the shape by simply overwriting the shape attribute of the object!
## Note: if you want to run this cell a second time, you need to run the previous
## cell also again before you do. Why? 

a.shape=(2,12)
print('a reshaped into a 2x12 matrix','\n',a)

## we may also read through the elements of the array by using an iterator "np.nditer". 
## You should read it as n-d iter(ator). In the following, python feeds into x each
## successive element of the n-d array. In C or Matlab, you had to write i,j,k as in 
## in the above loop---python gives you a quick way to do exactly the same thing.

print('Elements of the matrix returned in order from nditer:')
for x in np.nditer(a):
    print(x)

    
## we may also alter the array by using "op_flags" with "np.nditer"
for x in np.nditer(a, op_flags = ["readwrite"]): 
    x[...] = x*x


## We see the results of individually squaring each element
print("Here we calculate individually squaring the elements of a, resulting in \n ",a)

[[[ 0.  1.  2.  3.]
  [ 4.  5.  6.  7.]
  [ 8.  9. 10. 11.]]

 [[12. 13. 14. 15.]
  [16. 17. 18. 19.]
  [20. 21. 22. 23.]]]
a[0]=
 [[ 0.  1.  2.  3.]
 [ 4.  5.  6.  7.]
 [ 8.  9. 10. 11.]]
a[:,0,:]= 
 [[ 0.  1.  2.  3.]
 [12. 13. 14. 15.]]
Locations where the entry is <10 
 (array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), array([0, 0, 0, 0, 1, 1, 1, 1, 2, 2]), array([0, 1, 2, 3, 0, 1, 2, 3, 0, 1]))
Entries of a that are < 10 
 [0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
a reshaped into a 2x12 matrix 
 [[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11.]
 [12. 13. 14. 15. 16. 17. 18. 19. 20. 21. 22. 23.]]
Elements of the matrix returned in order from nditer:
0.0
1.0
2.0
3.0
4.0
5.0
6.0
7.0
8.0
9.0
10.0
11.0
12.0
13.0
14.0
15.0
16.0
17.0
18.0
19.0
20.0
21.0
22.0
23.0
Here we calculate individually squaring the elements of a, resulting in 
  [[  0.   1.   4.   9.  16.  25.  36.  49.  64.  81. 100. 121.]
 [144. 169. 196. 225. 256. 289. 324. 361. 400. 441. 484. 529.]]


## <span style="color:red"> (3) Questions</span>

1. Create a new code cell for each problem below and type in the code.

2. Create a list with at least the names of four Hawaiian Islands. If desired you can write the ʻokina 
as `u"\u02BB" `. For example to write hawai'i, we would type,
```python
okina=u"\u02BB"
a="hawai" + okina + "i"
```

In [23]:
okina=u"\u02BB"
Hawaii = "Hawai" + okina + "i"
Oahu = "O" + okina + "ahu"
Kauai = "Kaua" + okina + "i"
Molokai = "Moloka" + okina + "i"
islands = [Hawaii, Oahu, Kauai, Molokai]

3. Iterate through the list and print the contents.

In [24]:
for x in islands:
    print(x)

Hawaiʻi
Oʻahu
Kauaʻi
Molokaʻi


4. Create a tuple containing three elements: the integers 3, 4, and 2 in that order.

In [25]:
tuple = (3, 4, 2)
print(tuple)

(3, 4, 2)


5. Create a numpy three dimensional array of 64 bit floats with the shape from the previous problem, and initialize with zeros.

In [26]:
#numpy.zeros(shape, dtype=float, order='C')
#Whether to store multi-dimensional data in row-major (C-style) or column-major (Fortran-style) order in memory.
numpy = np.zeros(tuple, dtype = np.float64)

# To see if we have the correct shape)
print("Shape:", numpy.shape)
# Print if populated with zeros
print("Array:\n", numpy)

Shape: (3, 4, 2)
Array:
 [[[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]
  [0. 0.]]]


6. Set the 0,0,0th entry of the matrix to be 1/3. Now iterate through the array using three `for` loops. In each successive location from the 0,0,1th location, write the square root of the number from the previous location. Repeat the same with np.nditer.

In [27]:
################
# USING 3 FOR LOOPS
################

############
# Variables
############
# First set 0,0,0th entry of the matrix to be 1/3
# In this case, my matrix is called "numpy"
numpy[0,0,0] = 1/3

# We will create a previous variable to keep on track of the number from previous location
# Since we start at 0,0,1th position, the value of previous should be numpy[0,0,0]
previous = numpy[0,0,0]

##############
# Loops
#############
#First Loop, will loop 3 times since numpy.shape = [3,4,2]
#Thus, numpy.shape = [i,j,k]
for i in range(numpy.shape[0]):
    # Second loop, will loop 4 times
    for j in range(numpy.shape[1]):
        #Third loop, will loop 2 times
        for k in range(numpy.shape[2]):
            # From the 0,0,1th location, do te sqrt root
            # Thus, if it is 0,0,0th location; do nothing
            if (i == 0 and j == 0 and k == 0):
                continue
            # Do sqrt root
            else:
                numpy[i,j,k] = np.sqrt(previous)
                previous = numpy[i,j,k]
##############
# Print result 
##############
print("Result:\n",numpy)

Result:
 [[[0.33333333 0.57735027]
  [0.75983569 0.87168554]
  [0.93364101 0.96625101]
  [0.98298068 0.99145382]]

 [[0.99571774 0.99785657]
  [0.99892771 0.99946371]
  [0.99973182 0.9998659 ]
  [0.99993295 0.99996647]]

 [[0.99998324 0.99999162]
  [0.99999581 0.9999979 ]
  [0.99999895 0.99999948]
  [0.99999974 0.99999987]]]


In [28]:
################
# USING np.diter
################
#First we will empty the numpy array since we populated the values form the previous step.
numpy = numpy*0

############
# Variables
############
# First set 0,0,0th entry of the matrix to be 1/3
# In this case, my matrix is called "numpy"
numpy[0,0,0] = 1/3

# We will create a previous variable to keep on track of the number from previous location
# Since we start at 0,0,1th position, the value of previous should be numpy[0,0,0]
previous = numpy[0,0,0]

# A counter variable used to skip first entry
count = 0;
##############
# nditer
#############
# Since we want to change the value of numpy,
# We updated the defaulted flag of read only to write as well.
for i in np.nditer(numpy, op_flags=['readwrite']):
    # Skip first entry
    if count == 0:
        count = count + 1
    # populate numpy
    # i[...] means current cell/location
    else:
        i[...] = np.sqrt(previous)
        previous = i
    
##############
# Print result 
##############
print("Result:\n",numpy)

Result:
 [[[0.33333333 0.57735027]
  [0.75983569 0.87168554]
  [0.93364101 0.96625101]
  [0.98298068 0.99145382]]

 [[0.99571774 0.99785657]
  [0.99892771 0.99946371]
  [0.99973182 0.9998659 ]
  [0.99993295 0.99996647]]

 [[0.99998324 0.99999162]
  [0.99999581 0.9999979 ]
  [0.99999895 0.99999948]
  [0.99999974 0.99999987]]]


7. Iterate through the array using `np.nditer` and print out each value.

In [29]:
# Indexes
i = 0
j = 0
k = 0

for x in np.nditer(numpy, flags=['f_index']):
    # check if j need to increase
    if k == 2:
        j = j + 1
        k = 0
    # check if i need to increase
    if j == 4:
        i = i + 1
        j = 0
        k = 0
    #print values    
    print("numpy[",i,j,k,"]: ", x)
    # increase k
    k = k + 1

numpy[ 0 0 0 ]:  0.3333333333333333
numpy[ 0 0 1 ]:  0.5773502691896257
numpy[ 0 1 0 ]:  0.7598356856515925
numpy[ 0 1 1 ]:  0.8716855428717357
numpy[ 0 2 0 ]:  0.93364101391902
numpy[ 0 2 1 ]:  0.9662510097893922
numpy[ 0 3 0 ]:  0.9829806762034502
numpy[ 0 3 1 ]:  0.9914538195011657
numpy[ 1 0 0 ]:  0.9957177408789931
numpy[ 1 0 1 ]:  0.9978565733004885
numpy[ 1 1 0 ]:  0.9989277117491978
numpy[ 1 1 1 ]:  0.9994637120722282
numpy[ 1 2 0 ]:  0.9997318200758782
numpy[ 1 2 1 ]:  0.9998659010466745
numpy[ 1 3 0 ]:  0.9999329482753704
numpy[ 1 3 1 ]:  0.9999664735756746
numpy[ 2 0 0 ]:  0.9999832366473324
numpy[ 2 0 1 ]:  0.9999916182885397
numpy[ 2 1 0 ]:  0.9999958091354881
numpy[ 2 1 1 ]:  0.9999979045655486
numpy[ 2 2 0 ]:  0.9999989522822255
numpy[ 2 2 1 ]:  0.9999994761409755
numpy[ 2 3 0 ]:  0.9999997380704535
numpy[ 2 3 1 ]:  0.9999998690352182


8. Write a function to multiply a matrix with a vector on the right. Do this with a loop and using nditer. In the next lab, we will see a faster way to compute this.

For example to write a function to print the elements of a matrix using a loop, we write
```python
def print_matrix(x):
    if len(x.shape) != 2:
        print("x must be a two-dimensional array.")
        return None
    for i in range(x.shape[0]):
        for j in range(x.shape[1]):
            print(x[i,j]," ", end = '')
        print("\n")
```

In [30]:
##########################
# LOOP
##########################
def For(matrix, vector, type):
    row, col = matrix.shape
    vrow = vector.shape[0]

    if (col != vrow):
        return "'Sorry; your matrix's column must be equal to your vectors' row."
        
    # Create new matrix body
    new = np.zeros((row, 1), type)
    
    # Populate new matrix
    for i in range(row):
        for j in range(col):
            # Multiplication
            new[i][0] += matrix[i][j] * vector[j]
            
    return new

##############
# TEST
##############
print("-------------Loop---------------")
print("VALID INPUT FOR FOR LOOP:\n",For(np.array([[1, 1], [1, 1]]), np.array([1, 1]), np.int)) # valid mult
print("INVALID INPUT FOR FOR LOOP:\n",For(np.array([[1, 1], [1, 1]]), np.array([1, 1, 1]), np.int)) # invalid mult

def NDITER(matrix, vector, type):
    row, col = matrix.shape
    vrow = vector.shape[0]
    
    if (col != vrow):
        return "'Sorry; your matrix's column must be equal to your vectors' row."

    new = np.zeros((row, 1), type)
    
    i = 0
    j = 0

    for x in np.nditer(matrix, op_flags = ["readwrite"]):
        new[i][0] += x * vector[j]
        j += 1
        if (j == col):
            j = 0
            i += 1

    return new

##############
# TEST
##############
print("-------------NDITER---------------")
print("VALID INPUT FOR FOR LOOP:\n",For(np.array([[1, 1], [1, 1]]), np.array([1, 1]), np.int)) # valid mult
print("INVALID INPUT FOR FOR LOOP:\n",For(np.array([[1, 1], [1, 1]]), np.array([1, 1, 1]), np.int)) # invalid mult


-------------Loop---------------
VALID INPUT FOR FOR LOOP:
 [[2]
 [2]]
INVALID INPUT FOR FOR LOOP:
 'Sorry; your matrix's column must be equal to your vectors' row.
-------------NDITER---------------
VALID INPUT FOR FOR LOOP:
 [[2]
 [2]]
INVALID INPUT FOR FOR LOOP:
 'Sorry; your matrix's column must be equal to your vectors' row.


9. In a 2-d array (matrix), write a function that returns the diagonal elements as an array. Ask your TA to illustrate the diagonal elements for a rectangular array before you begin. 

In [31]:
#########################################
# Diagonal elements:
# -----------------------
# Example 1:
# -----------------------
# 1 2 3
# 4 5 6
# 7 8 9
# Return:
# [1,5,9]
# -----------------------
# Example 2:
# -----------------------
# 1 2 3
# 4 5 6
# Return:
# [1,5]
#########################################
import numpy as np # Just in case if this is the first cell you run
###############
# Function
###############
def diagonal(matrix, type):
    #matrix.shape returns row and column
    # so row = row; col = col; 
    row, col = matrix.shape

    limit = 0
    is_col = 0
    diagonal = 0

    ####################################################################################
    # Rectangular Matrix and Square Matrix check case:
    ####################################################################################
    # Check if we loop by col
    # row == col; is when square matrix; it does'nt matter if loop by col or row.
    if row > col or row == col:
        limit = col
        is_col = 1
        # populate 0 array
        diagonal = np.zeros(col, type)
    # Check if we loop by row
    elif row < col:
        limit = row
        is_col = 0
        # populate 0 array
        diagonal = np.zeros(row, type)
    

    j = 0
    for i in range(limit):
        # If col
        if is_col == 1:
            diagonal[i] = matrix[i][j]
        # If row
        else:
            diagonal[i] = matrix[j][i]
        j += 1
        
    return diagonal

##################
# Test 1
##################
print("============Test============")
print("Array::::\n",np.array([[1, 2],[3, 4]]))
print("Diagonal::::\n",diagonal(np.array([[1, 2], [3, 4]]), np.int))
##################
# Test 2
##################
print("============Test============")
print("Array::::\n",np.array([[5, 6], [7, 8], [9, 10]]))
print("Diagonal::::\n",diagonal(np.array([[5, 6], [7, 8], [9, 10]]), np.int))
##################
# Test 3
##################
print("============Test============")
print("Array::::\n",np.array([[11, 12, 13], [14, 15, 16]]))
print("Diagonal::::\n",diagonal(np.array([[11, 12, 13], [14, 15, 16]]), np.int))

============Test============
Array::::
 [[1 2]
 [3 4]]
Diagonal::::
 [1 4]
============Test============
Array::::
 [[ 5  6]
 [ 7  8]
 [ 9 10]]
Diagonal::::
 [5 8]
============Test============
Array::::
 [[11 12 13]
 [14 15 16]]
Diagonal::::
 [11 15]


10. Populate a 2-d array of shape (10,15) with 0s and 1s chosen randomly. To do this, import the module random first, and use randint(0,1) to generate a random bit. Use nditer instead of a loop here. 

In [32]:
# since we want to populate randomly, import the random library
import random

# populate a 10x15 array with numbers
rand_array = np.zeros((10, 15), np.int)

# Now populate the array with 0 and 1 using np.nditer
# don't forget the ReadWrite since we want to update the 0 that were populated.
for x in np.nditer(rand_array, op_flags = ["readwrite"]): 
    x[...] = random.randint(0, 1)

#Display the output
print("Output:::\n", rand_array)

Output:::
 [[1 1 1 0 0 0 0 0 1 0 1 0 0 0 1]
 [1 1 1 0 0 0 1 1 1 1 1 0 0 1 0]
 [1 1 1 0 0 0 1 0 0 1 0 1 0 0 1]
 [0 1 0 1 1 0 0 1 0 1 1 1 0 0 1]
 [0 0 1 1 0 0 0 1 1 1 1 0 1 0 0]
 [1 1 1 1 0 0 1 1 0 1 0 0 1 1 1]
 [0 0 1 1 1 0 1 0 1 0 0 0 0 1 0]
 [0 1 0 1 1 0 0 0 0 0 1 1 1 1 0]
 [0 0 0 1 1 1 0 0 0 1 1 0 0 0 1]
 [1 0 0 0 0 0 1 0 1 0 1 0 0 0 0]]


11. You can reshape arrays using the call np.reshape(a, newshape), where newshape is the tuple corresponding to the new shape desired. Generate an array of 24 increasing numbers starting from 0 using range(24). Reshape it to get the 3-d matrix a. What happens when the tuple newshape is incompatible with the size of a---ie, here a has size 24, but what if newshape is (5,5)?

In [33]:
# Creating new array of 24.
new_array = np.zeros(24, np.int)

# Now populate the array in increasing order; so 0, 1,2,3, ..., 24
for i in range(24):
    new_array[i] = i

# To check if it works
print("1D::::\n", new_array)

###############
## Reshaping
##############
# 3d Array
ThreeDArray = new_array.reshape(3,4,2)
print("3D:::::\n", ThreeDArray)

# What if it is incompatible?
# FailArray = new_array.reshape(5,5)
# Returns: ValueError: cannot reshape array of size 24 into shape (5,5)

1D::::
 [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]
3D:::::
 [[[ 0  1]
  [ 2  3]
  [ 4  5]
  [ 6  7]]

 [[ 8  9]
  [10 11]
  [12 13]
  [14 15]]

 [[16 17]
  [18 19]
  [20 21]
  [22 23]]]


12. The function np.where(condition, x, y) evaluates the condition. x,y are matrices of the same size. If condition is TRUE in a given position, it outputs the corresponding element of x in that position, else y. Take the matrix from Problem 10 above, and replace every occurence of 0 with -1 using np.where(condition, x, y) by choosing the arguments wisely.

In [34]:
####################
# How it works
###################
# np.where(condition, if true, if false)
new_rand_array = np.where(rand_array == 0, -1,rand_array)
print(new_rand_array)

[[ 1  1  1 -1 -1 -1 -1 -1  1 -1  1 -1 -1 -1  1]
 [ 1  1  1 -1 -1 -1  1  1  1  1  1 -1 -1  1 -1]
 [ 1  1  1 -1 -1 -1  1 -1 -1  1 -1  1 -1 -1  1]
 [-1  1 -1  1  1 -1 -1  1 -1  1  1  1 -1 -1  1]
 [-1 -1  1  1 -1 -1 -1  1  1  1  1 -1  1 -1 -1]
 [ 1  1  1  1 -1 -1  1  1 -1  1 -1 -1  1  1  1]
 [-1 -1  1  1  1 -1  1 -1  1 -1 -1 -1 -1  1 -1]
 [-1  1 -1  1  1 -1 -1 -1 -1 -1  1  1  1  1 -1]
 [-1 -1 -1  1  1  1 -1 -1 -1  1  1 -1 -1 -1  1]
 [ 1 -1 -1 -1 -1 -1  1 -1  1 -1  1 -1 -1 -1 -1]]
